File: cucumber-expressions/java/src/main/java/io/cucumber/cucumberexpressions/GeneratedExpression.java
package io.cucumber.cucumberexpressions;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class GeneratedExpression {
    private final String expressionTemplate;
    private final List<ParameterType<?>> parameterTypes;

    public GeneratedExpression(String expressionTemplate, List<ParameterType<?>> parameterTypes) {
        this.expressionTemplate = expressionTemplate;
        this.parameterTypes = parameterTypes;
    }

    public String getSource() {
        List<String> parameterTypeNames = new ArrayList<>();
        for (ParameterType<?> parameterType : parameterTypes) {
            String name = parameterType.getName();
            parameterTypeNames.add(name);
        }
        return String.format(expressionTemplate, parameterTypeNames.toArray());
    }

    /**
     * @deprecated use {@link #getParameterNames(LetterCase, ProgrammingLanguage)}
     */
    @Deprecated
    public List<String> getParameterNames() {
        return getParameterNames(LetterCase.LOWER_CAMEL_CASE, ProgrammingLanguage.JAVA);
    }

    public List<String> getParameterNames(LetterCase letterCase, ProgrammingLanguage programmingLanguage) {
        HashMap<String, Integer> usageByTypeName = new HashMap<>();
        List<String> list = new ArrayList<>();
        for (ParameterType<?> parameterType : parameterTypes) {
            String parameterName = getParameterName(parameterType.getName(), usageByTypeName, letterCase, programmingLanguage);
            list.add(parameterName);
        }
        return list;
    }

    public List<ParameterType<?>> getParameterTypes() {
        return parameterTypes;
    }

    private String getParameterName(String typeName, Map<String, Integer> usageByTypeName, LetterCase letterCase, ProgrammingLanguage programmingLanguage) {
        String name = letterCase.convert(typeName);

        Integer count = usageByTypeName.get(name);
        count = count != null ? count + 1 : 1;
        usageByTypeName.put(name, count);

        return count == 1 && !programmingLanguage.isKeyword(name) ? name : name + count;
    }
}

File: cucumber-expressions/go/generated_expression.go
package cucumberexpressions

import "fmt"

type GeneratedExpression struct {
	expressionTemplate string
	parameterTypes     []*ParameterType
}

func NewGeneratedExpression(expressionTemplate string, parameterTypes []*ParameterType) *GeneratedExpression {
	return &GeneratedExpression{
		expressionTemplate: expressionTemplate,
		parameterTypes:     parameterTypes,
	}
}

func (g *GeneratedExpression) Source() string {
	names := make([]interface{}, len(g.parameterTypes))
	for i, p := range g.parameterTypes {
		names[i] = p.Name()
	}
	return fmt.Sprintf(g.expressionTemplate, names...)
}

func (g *GeneratedExpression) ParameterNames() []string {
	usageByTypeName := map[string]int{}
	result := make([]string, len(g.parameterTypes))
	for i, p := range g.parameterTypes {
		result[i] = getParameterName(p.Name(), usageByTypeName)
	}
	return result
}

func (g *GeneratedExpression) ParameterTypes() []*ParameterType {
	return g.parameterTypes
}

func getParameterName(typeName string, usageByTypeName map[string]int) string {
	count, ok := usageByTypeName[typeName]
	if !ok {
		count = 0
	}
	usageByTypeName[typeName] = count + 1
	if count == 1 {
		return typeName
	}
	return fmt.Sprintf("%s%d", typeName, count)
}
